[TECH] Réduire le pic mémoire lors du parsing des fichiers SIECLE volumineux et optimiser le traitement (Pix-1405)

yahyayouness opened the pull request:
## :unicorn: Problème
L'import des fichiers SIECLE volumineux nécessite beaucoup de mémoire à savoir: la taille du fichier multiplié par 13.
Ceci est souvent la cause des crashs de conteneurs en production qui sont limités à 256 MB de mémoire.

En effet, en analysant la consommation mémoire du traitement de l'import, on observe un pic lors du parsing du fichier, qui s'accentue de plus en plus en fonction de sa taille. Ceci risque de générer la stacktrace suivante liée à la saturation mémoire:
`FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

![image](https://user-images.githubusercontent.com/10045497/97238205-1f9af580-17e9-11eb-9860-0be13b6584e6.png)

Le pic est engendré lors de la transformation du contenu du fichier SIECLE en un objet DOM.
`return new DOMParser().parseFromString(stringifiedXml);`

## :robot: Solution

Optimisation en 5 étapes:

**Etape 1**: Trouver une représentation moins couteuse que DOM en terme d'allocation mémoire.
Ci-dessous un benchmark des librairies et une comparaison des représentations JS , XML et JSON.

![image](https://user-images.githubusercontent.com/10045497/97238733-63422f00-17ea-11eb-9e59-ecb88b5e52e0.png)

On voit bien que la répresentation JS et JSON sont moins couteuses que celle du DOM générée par la lib DOMParser. 
De ce fait, elle sera remplacée par une représentation en JS. On constate donc une nette amélioration vu que cette dernière consomme 9 fois moins la taille du fichier. Toutefois, le problème du pic n'est toujours pas résolu.

**Etape 2:** 
- Opter pour le mode fichier de Hapi js qui permet de sauvegarder le contenu dans un fichier temporaire. Sur ce dernier, on va streamer noeud par noeud et extraire les données progressivement. 
- En combinant cette méthode avec l'optimisation décrite lors de l'étape 1 , le pic est résolu et laisse place à une consommation constante de la mémoire qui plus est se libère rapidement.  (voir graph ci-dessous)

![image](https://user-images.githubusercontent.com/10045497/97239362-b4065780-17eb-11eb-8d80-941d2f70efb5.png)

Exemple pour un fichier de 10Mb:
![image](https://user-images.githubusercontent.com/10045497/97885398-996e3a00-1d27-11eb-8e13-cafeb9d131ec.png)

**Etape 3:** Le streaming du fichier va permettre de:
-  déporter certaines vérifications  des données au début du traitement , évitant ainsi de parser entièrement un fichier voué à l'échec.
- filtrer des données au fur et à mesure  au lieu de faire le filtre après le parsing (Exemple pour le check de l'UAI de l'organisation qui permet de lever une erreur au premier parsing du premier noeud)

**Etape 4:** Faire un traitement par lot pour la partie insertion en base de donnée des schoolings registrations. (Faite dans cette PR)

**Etape 5:** Faire un streaming du fichier directement sans copie temporaire du fichier (Sera faite dans une autre PR)


## :rainbow: Remarques
Le traitement de parsing s'arrête à la première erreur ce qui représente un gain important au niveau temps de réponse et de consommation mémoire.
La consolidation de la consommation mémoire a été faite en se basant sur 3 fichiers de 1.9Mb, 5.5Mb, 10,2Mb
Le détail de la consommation ici https://docs.google.com/spreadsheets/d/1TLs6tkzTyUHZD1wPbSL1nRjHD0EZoM-kwcg_7-DmQ98/edit#gid=1246792855

Des tests sont en cours de finalisation avec une reflexion autour la gestion fs dans les tests. Mon mock fs n'a pas bien marché, d'ou la création de fichiers temporaires pour les tests. 
Une optimisation de l'import csv a été faite en remplaçant le buffer hapi par un buffer depuis le fichier qui sera rapidement purgé vu que la portée de la variable est locale.


## :100: Pour tester
⚠️  Vu les changements structurants, des tests de non régression sont à prévoir pour toute la fonctionnalité import csv et xml.

Commit:
[API] install sax, saxpath, xml2js libs

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2061.review.pix.fr
- Orga: https://orga-pr2061.review.pix.fr
- Certif: https://certif-pr2061.review.pix.fr
- Admin: https://admin-pr2061.review.pix.fr
- API: https://api-pr2061.review.pix.fr/api/

 Please check it out!

Commit:
[API] change hapi output mode to file for upload

The output mode file write the incoming file in a temporary one before doing further work.
The parse property determines if the incoming payload gets parsed (true, default) or presented raw (false).

Commit:
[API] use payload with file path at controller

Commit:
[API] manage purge tmp siecle file after parsing

remove check uai from siecle in usecase.

Commit:
[API] use buffer from file for siecle csv file

Instead to use of global buffer of hapi that use memory we create a temporary buffer from file that will be removed quickly

Commit:
[API] add siecle files inputs for tests

Commit:
[API] stream file and transorm xml to js instead of DOM

Remove DOMparser that allow a lot of memory.
Stream file and validate functionnal rule after parsing each element.
Start with filtering data after each chunk.
Throw an error if duplicate student.

yahyayouness commented on api/tests/acceptance/application/organizations/organization-controller-import-schooling-registrations_test.js:
```diff
@@ -4,6 +4,7 @@ const {
   expect, knex, databaseBuilder,
   generateValidRequestAuthorizationHeader,
 } = require('../../../test-helper');
+require('events').EventEmitter.defaultMaxListeners = 60;
```
J'ai enlevé les listeners après le traitement de streaming, mais j'ai toujours un warning sur ce test d'ou l'augmentation du maxListener. Je creuse encore pour pouvoir le purger.

    fileStream.on('end', () => {

      saxParser.removeAllListeners();
      streamerToParseOrganizationUAI.removeAllListeners();
      streamerToParseSchoolingRegistrations.removeAllListeners();

    });

yahyayouness replied:
Résolu avec Jonathan, le saxParser était partagé par les deux streamers. Du coup, les listeners s'enrichissaient à fur à mesure avant de se libérer à la fin. 
Fix: Déclarer deux parseurs comme variables locales, faire successivement le premier parsing de l'UAI. Si ce dernier est valide, on ouvre un deuxième stream pour parser la suite. 
En cas d'erreur, on reject avec une suppression des listeners et  une destruction du stream dans le bloc finaly.

octo-topi commented:
> Des tests sont en cours de finalisation avec une reflexion autour la gestion fs dans les tests. 

Ok donc actuellement, comment on teste  la perf sur des fichiers de plus en plus gros ?
On les crée à la main ?

Commit:
[API] add temporary limit for max size request

yahyayouness commented:
> > Des tests sont en cours de finalisation avec une reflexion autour la gestion fs dans les tests.
> 
> Ok donc actuellement, comment on teste la perf sur des fichiers de plus en plus gros ?
> On les crée à la main ?

On dispose de 3 fichiers ( 1.9Mb , 5.5Mb et 10,2Mb ) 
La consommation des 3 fichiers sont constantes au niveau du parsing du fichier.
Voir https://docs.google.com/spreadsheets/d/1TLs6tkzTyUHZD1wPbSL1nRjHD0EZoM-kwcg_7-DmQ98/edit#gid=1246792855 

Pour les JDD, je vais te les fournir .

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+  const firstline = await _readFirstLineFromFile(payload.path);
+  const encoding = xmlEncoding(Buffer.from(firstline)) || DEFAULT_FILE_ENCODING;
+
+  return new Promise(function(resolve, reject) {
```
Pourquoi passer par une écriture Promise à l'ancienne ? Le standard depuis un certain temps est de privilégier autant que possible la forme `async` / `await`. En l'occurrence, `reject` n'apparaît que 2 fois. Je pense qu'il est très facile de faire 1 seul `try/catch` avec 2 instruction `throw new Error()` en cas d'erreur.

yahyayouness replied:
Merci pour ton retour. 

Async / Await fonctionne bien avec des promesses. Mais avec les streams ( createReadStream ) c'est moins évident. D'ou le wrapping vers une promesse pour faire un await dessus.

Voilà ce qui a été testé en premier:

1. const fs = require('fs').promises mais ca ne marche pas avec le createReadStream car ca renvoie un stream. 
Avec un readFile() ca renvoie bien une promesse.

2. Une option de lire le stream sous forme de promise: ( à l'image de la fonction streamToPromise ci-dessous ) 
Cette option a été retenu en faisant du traitement par chunk. 

```
function streamToPromise(stream) {
  return new Promise((resolve, reject) => {
    let totalData = '';
    stream.on('data', (data) => {
      totalData += data;
    });
    stream.on('end', () => {
      resolve(totalData);
    });
    stream.on('error', reject);
  });
} 
```
N'hésitez pas à faire des propositions, je suis preneur pour trouver mieux. poke @jonathanperret

yahyayouness replied:
J'ai pris en compte ta remarque @jonathanperret et utiliser la version xml2js qui renvoie une promesse.

yahyayouness replied:
Rollback vers la version xml2js qui utilise une callback car elle est plus optimisée du fait qu'elle parse les blocs successivement et non de facon asynchrone.
Utilisation la fonction undefined de lodash pour expliciter le filtre.

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+    const mapSchoolingRegistrationsByStudentId = new Map();
+    const nationalStudentIds = [];
+
+    const handlerParseOrganizationUAI = function(xmlFormat) {
```
Pour être consitant avec le reste du code, j'utiliserais la notation _fat arrow_.

yahyayouness replied:
Thanks, je les avais sorti en fonction pour pouvoir debug la suppression des events listeners à la fin. 
streamerToParseOrganizationUAI. removeListener(handlerXXX);

Je les ai remis directement en fat arrow sans les variables et en utilisant le removeAllListeners.

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+    };
+    streamerToParseOrganizationUAI.on('match', handlerParseOrganizationUAI);
+
+    const handlerParseSchoolingRegistrations = function(xmlFormat) {
```
Idem, je privilégierais la notation _fat arrow_.

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+                }
+              });
+
+              if (currentStudent.division === undefined) {
```
Privilégier le shortcut `if (!currentStudent.division)` qui tient en plus compte du cas de `null`

yahyayouness replied:
Justement, je cherche à verifier seulement si la propriété est undefined. 
Si la variable est null ou à empty suite à une balise <CODE_STRUCTURE><CODE_STRUCTURE>, je dois pas supprimer l'élève.

jbuget replied:
Ah ! ok. Dans ce cas, j'utiliserais peut-être une émthode lodash ou à nous spécifique qui met bien en évidence qu'on veut tester vraiment `undefined`. Car en l'état, si je débarque dans le code, j'aurais introduis une régression en voulant bien faire.

yahyayouness replied:
J'avoue, je vais créer une fonction qui explicite ce check fonctionnellement et qui met en evidence juste l'undefined.

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+
+    const resolveSchoolingRegistrations = function() {
+      resolve(Array.from(mapSchoolingRegistrationsByStudentId.values()).filter((schoolingRegistration) => schoolingRegistration.division !== undefined));
+    };
```
pour vérifier l'existence d'une valeur, utiliser le shortcut `!!schoolingRegistration.division ` plutôt que `!== undefined` qui ne tient pas compte de la _valeur_ `null`

yahyayouness replied:
Voir commentaire plus haut

jbuget commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,151 @@
+      }
+    })
+      .on('close', () => resolve(value.slice(value.charCodeAt(0) === BOM ? 1 : 0, position)))
+      .on('error', (err) => reject(err));
```
si on conserce l'écriture en promesses, un simple 

```
.on('error', reject);
```

fait l'affaire :

Commit:
[API] chunk insertion schooling registration db

In order to serve a large file and do not use a lot of memory, we chunk the insertion in database by 1000 by default

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
+    const fileStream = fs.createReadStream(payload.path).pipe(iconv.decodeStream(encoding));
+
+    const streamerToParseOrganizationUAI = new saxPath.SaXPath(saxParser, NODE_ORGANIZATION_UAI);
+    const streamerToParseSchoolingRegistrations = new saxPath.SaXPath(saxParser, NODES_SCHOOLING_RESGISTRATIONS);
```
```suggestion
    const streamerToParseSchoolingRegistrations = new saxPath.SaXPath(saxParser, NODES_SCHOOLING_REGISTRATIONS);
```

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
+const DEFAULT_FILE_ENCODING = 'iso-8859-15';
+const DIVISION = 'D';
+const NODE_ORGANIZATION_UAI = '/BEE_ELEVES/PARAMETRES';
+const NODES_SCHOOLING_RESGISTRATIONS = '/BEE_ELEVES/DONNEES/*/*';
```
```suggestion
const NODES_SCHOOLING_REGISTRATIONS = '/BEE_ELEVES/DONNEES/*/*';
```

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
-  return { UAIFromSIECLE, resultFromExtraction };
-}
+    streamerToParseSchoolingRegistrations.on('match', (xmlFormat) => {
+      if (xmlFormat.startsWith(ELEVE_ELEMENT) || xmlFormat.startsWith(STRUCTURE_ELEVE_ELEMENT)) {
```
C'est une question de préférence, mais j'aime bien essayer de rendre le code un poil plus expressif.
Déjà, je renommerais `xmlFormat` en `xmlNode`, puis je ferais :
```suggestion
    streamerToParseSchoolingRegistrations.on('match', (xmlNode) => {
      if (_isSchoolingRegistrationNode(xmlNode)) {
      /*...*/
      }
   }
  
  function _isSchoolingRegistrationNode(xmlNode) {
    return xmlNode.startsWith(ELEVE_ELEMENT) || xmlNode.startsWith(STRUCTURE_ELEVE_ELEMENT)
  }
```

yahyayouness replied:
Done, j'aime bien aussi

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
-}
+    streamerToParseSchoolingRegistrations.on('match', (xmlFormat) => {
+      if (xmlFormat.startsWith(ELEVE_ELEMENT) || xmlFormat.startsWith(STRUCTURE_ELEVE_ELEMENT)) {
+        parseString(xmlFormat, function(err, jsFormat) {
```
Je trouve pas jsFormat très joli comme terme
```suggestion
        parseString(xmlNode, function(err, nodeData) {
```

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
+        parseString(xmlFormat, function(err, jsFormat) {
+
+          if (jsFormat.ELEVE) {
+            const isStudentNotLeftSchoolingRegistration = isEmpty(jsFormat.ELEVE.DATE_SORTIE);
+            const isStudentNotYetArrivedSchoolingRegistration = !isEmpty(jsFormat.ELEVE.ID_NATIONAL);
+            const isStudentNotDuplicatedInTheSIECLEFile = !mapSchoolingRegistrationsByStudentId.has(jsFormat.ELEVE.$.ELEVE_ID);
+
+            if (isStudentNotLeftSchoolingRegistration && isStudentNotYetArrivedSchoolingRegistration && isStudentNotDuplicatedInTheSIECLEFile) {
```
```suggestion
            const studentData = jsFormat.ELEVE;
            if(_isStudentEligible(studentData, mapSchoolingRegistrationsByStudentId)) {
            /* ... */
            }
          }
/* ... */
function _isStudentEligible(studentData, mapSchoolingRegistrationsByStudentId) {
  const isStudentNotLeftSchoolingRegistration = isEmpty(studentData.DATE_SORTIE);
  const isStudentNotYetArrivedSchoolingRegistration = !isEmpty(studentData.ID_NATIONAL);
  const isStudentNotDuplicatedInTheSIECLEFile = !mapSchoolingRegistrationsByStudentId.has(studentData.$.ELEVE_ID);

  return isStudentNotLeftSchoolingRegistration && isStudentNotYetArrivedSchoolingRegistration && isStudentNotDuplicatedInTheSIECLEFile;
}
```

yahyayouness replied:
Done

laura-bergoens commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,147 @@
+
+          if (jsFormat.STRUCTURES_ELEVE && mapSchoolingRegistrationsByStudentId.has(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID)) {
+            const currentStudent = mapSchoolingRegistrationsByStudentId.get(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
+            const structureElement = jsFormat.STRUCTURES_ELEVE.STRUCTURE;
+            if (!isEmpty(structureElement)) {
+              structureElement.forEach((structure) => {
+                if (structure.TYPE_STRUCTURE[0] === DIVISION && structure.CODE_STRUCTURE[0] !== 'Inactifs') {
+                  currentStudent.division = structure.CODE_STRUCTURE[0];
+                }
+              });
+
+              if (currentStudent.division === undefined) {
+                mapSchoolingRegistrationsByStudentId.delete(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
+              }
+            }
```
Lodash pourrait t'éviter de vérifier que les collections ne sont pas vides, et donc réduire le nesting d'un cran
```suggestion
         const structureElement = jsFormat.STRUCTURES_ELEVE.STRUCTURE;
         _.each(structureElement, (structure) => {
                if (structure.TYPE_STRUCTURE[0] === DIVISION && structure.CODE_STRUCTURE[0] !== 'Inactifs') {
                  currentStudent.division = structure.CODE_STRUCTURE[0];
                }
         });
        if (currentStudent.division === undefined) {
            mapSchoolingRegistrationsByStudentId.delete(jsFormat.STRUCTURES_ELEVE.$.ELEVE_ID);
        }
```

yahyayouness replied:
Done

laura-bergoens commented:
Il est pas simple à lire ce service :grimacing:

laura-bergoens commented:
Je confirme la non régression de l'import CSV SCO-AGRI

octo-topi commented:
# Tests manuels => OK !

## Avant optimisation

**Résumé**
Si on importe
- un fichier de 10Mb
- sur une API dont la mémoire est limité à 250Mb comme Scalingo
`NODE_OPTIONS="--max-old-space-size=250" npm run start`
- cela utilise en mémoire 10 fois la taille du fichier (150 Mb)
- et cause le crash si d'autres requêtes sont traitées par Hapi (ce qui est le cas en production)

**Détail** 
150 Mb (81=>231) de mémoire sur un pic de < 10 ms sur l'étape de parsing du fichier reçu (avant réconciliation) [buildXMLDomFromBuffer](http://github.com/1024pix/pix/blob/dev/api/lib/domain/services/schooling-registrations-xml-service.js#L23-L23) dans extractSchoolingRegistrationsInformationFromSIECLE
```
enter _buildXmlDomFromBuffer
time--mm:ss:SSS;heap_mb
35:15:416;81
35:17:102;231
leave _buildXmlDomFromBuffer
```

Cet espace est réclamé presque instantanément par le GC
```
time--mm:ss:SSS;heap_mb
35:18:855;231
35:18:955;235
35:19:134;83
35:22:511;127
35:23:411;124
35:23:512;129
35:23:611;118
35:23:712;123
leave addOrUpdateOrganizationSchoolingRegistrations
leave importSchoolingRegistrationsFromSIECLE
```

Si on baisse la mémoire en dessous de 235Mb, l'application crashe
Comme en production, la mémoire est partagée avec d'autres requêtes, le conteneur crashera même au réglage acuel de 250Mb


## Après optimisation

**Résumé**

Si on importe
- un fichier de 10Mb
- sur une API dont la mémoire est limité à 250Mb comme Scalingo
`NODE_OPTIONS="--max-old-space-size=250" npm run start`
- cela utilise en mémoire 5 fois la taille du fichier (50 Mb), soit 2 fois moins qu'auparavant
- cela enlève le pic initial

**Détail** 

16 Mb (80=>96) de mémoire sur l'étape ciblée précédement [extractSchoolingRegistrationsInformationFromSIECLE](http://github.com/1024pix/pix/blob/pix-1405-stream-siecle-file/api/lib/domain/usecases/import-schooling-registrations-from-siecle.js#L32-L32)
50 Mb (80=>130) de mémoire sur le reste (réconciliation proprement dite)

```
enter importSchoolingRegistrationsFromSIECLE
time;heap_mb
04:00:122;79
enter extractSchoolingRegistrationsInformationFromSIECLE
04:00:140;80
04:00:289;90
04:00:412;86
(..)
04:06:489;92
leave importSchoolingRegistrationsFromSIECLE
04:06:603;96
04:07:020;116
04:07:120;121
04:08:622;127
04:08:790;130
04:09:178;107
04:12:080;90
leave importSchoolingRegistrationsFromSIECLE
```

yahyayouness commented:
> Il est pas simple à lire ce service 😬

Je te propose de faire une passe avec la team pour décrire le traitement avec un aspect plus fonctionnel , j'ai tenté mais j'avoue j'ai pas toute la connaissance métier.

Typiquement la notion de division, inactifs ...   j'ai repris les terms existants en essayant de faire mieux mais c'est pas assez.

jonathanperret committed:
[API] schooling-registrations-xml-service refactoring

octo-topi commented on api/lib/application/organizations/index.js:
```diff
@@ -173,8 +173,8 @@ exports.register = async (server) => {
         payload: {
-          maxBytes: 1048576 * 10, // 10MB
-          parse: 'gunzip',
+          maxBytes: 1048576 * 20, // 20MB
```
Je propose de créer une variable d'environnement, cela permettra de s'adapter si besoin

Comme noté dans la PR #2109, un fichier SIECLE de 4000 élèves fait au moins 10 Mo (sans balise OPTION)
La plus grande organisation compte déjà 3640 membres

```
  id   | count 
-------+-------
  1562 |  3640
  5370 |  3581
  5592 |  3515

SELECT
     o.id, COUNT(1)
FROM
    "schooling-registrations" sr
         INNER JOIN organizations o on sr."organizationId" = o.id
GROUP BY o.id
ORDER BY COUNT(1) DESC
;
```

yahyayouness replied:
Effectivement, une variable d'env avec une valeur par défaut ( xMb ) peut éviter un déploiement si on veut changer à jour la valeur. Je vois avec la team si on veut garder la limite ou pas et je modifie dans le cas écheant.

octo-topi commented on api/lib/application/organizations/index.js:
```diff
@@ -173,8 +173,8 @@ exports.register = async (server) => {
         payload: {
-          maxBytes: 1048576 * 10, // 10MB
-          parse: 'gunzip',
+          maxBytes: 1048576 * 20, // 20MB
```
```suggestion
          const OneMb = 1048576
          maxBytes: 20 *OneMb,
```

octo-topi commented:
Super tes commentaires dans les commit, par exemple
https://github.com/1024pix/pix/pull/2061/commits/3be185a45d259d141c239fd25151997cc0c6fedc

```
The output mode file write the incoming file in a temporary one before doing further work.
The parse property determines if the incoming payload gets parsed (true, default) or presented raw (false).
```

octo-topi commented:
PI, côté Front, le message est cas de fichier > limite est Quelque chose s'est mal passé. Veuillez réessayer.
Si on voit qu'il y a fréquemment des erreurs, à l'avenir, ça vaudrait le coup de dire qqch comme "Merci de contacter le support"

![image](https://user-images.githubusercontent.com/56302715/98373945-460a2d80-2040-11eb-9c03-9ce8eed54ce2.png)

Thanks, on peut mettre un message spécifique pour dire que le taille du fichier est supérieur à la taille limite ( voir avec la team prescription ) pour donner du feedback sur cette erreur.

octo-topi commented on api/lib/domain/usecases/import-schooling-registrations-from-siecle.js:
```diff
@@ -1,31 +1,39 @@
 
-module.exports = async function importSchoolingRegistrationsFromSIECLE({ organizationId, buffer, format, organizationRepository, schoolingRegistrationsXmlService, schoolingRegistrationRepository }) {
+const NO_SCHOOLING_REGISTRATIONS_FOUND = 'Aucune inscription d‘élève n‘a pu être importée depuis ce fichier. Vérifiez que le fichier est conforme.';
+const FILE_FORMAT_NOT_VALID = 'Format de fichier non valide.';
```
```suggestion
const INVALID_FILE_FORMAT = 'Format de fichier non valide.';
```

octo-topi commented on api/tests/unit/domain/usecases/import-schooling-registrations-from-siecle_test.js:
```diff
@@ -115,17 +122,22 @@ describe('Unit | UseCase | import-schooling-registrations-from-siecle', () => {
     context('because the file is not valid', () => {
```
```suggestion
    context('because the file content is invalid', () => {
```

octo-topi commented on api/tests/unit/domain/usecases/import-schooling-registrations-from-siecle_test.js:
```diff
@@ -174,16 +169,21 @@ describe('Unit | UseCase | import-schooling-registrations-from-siecle', () => {
+
+        await fs.writeFile(filePath, '');
         // when
-        const result = await catchErr(importSchoolingRegistrationsFromSIECLEFormat)({ organizationId, buffer, format, organizationRepository: organizationRepositoryStub, schoolingRegistrationsXmlService: schoolingRegistrationsXmlServiceStub, schoolingRegistrationRepository: schoolingRegistrationRepositoryStub });
```
l.194
` context('because a nationalStudentId appears twice in the file', () => {`

octo-topi commented on api/tests/unit/domain/usecases/import-schooling-registrations-from-siecle_test.js:
```diff
@@ -136,34 +148,17 @@ describe('Unit | UseCase | import-schooling-registrations-from-siecle', () => {
         // then
         expect(result).to.be.instanceOf(FileValidationError);
-        expect(result.message).to.equal('Aucun étudiant n’a été importé. L’import n’est pas possible car l’UAI du fichier SIECLE ne correspond pas à celui de votre établissement. En cas de difficulté, contactez support.pix.fr.');
+        expect(result.message).to.equal('Format de fichier non valide.');
```
Au début je pensais que c'était dû au fait que les données n'étaient pas valides, et donc je me demande si c'était la même chose que 'because the file is not valid', mais en fait c'est que les données sont OK mais que le BDD les rejette
```suggestion
context('because information save has failed in database'', () => {
```

octo-topi commented:
Tu peux fixup le commit `[API] use buffer from file for siecle csv `file  dans  `[API] manage purge tmp siecle file after parsing ` car le nom n'est pas le bon (contenu réel = extraction d'un message dans une constante FILE_FORMAT_NOT_VALID)

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,167 @@
-const { DOMParser } = require('xmldom');
-const xmlBufferToString = require('xml-buffer-tostring');
+const { FileValidationError, SameNationalStudentIdInFileError } = require('../errors');
```
Description du commit

[API] stream file and transorm xml to js instead of DOM 
[API] parse XML using stream instead of AST

Remove DOM parser that allow a lot of memory.
Replace DOM parser that build the XML AST (thus allocating too much memory for a large file) by SAX parser that use streams (having a lightweight footprint)

octo-topi commented on api/lib/domain/usecases/import-schooling-registrations-from-siecle.js:
```diff
@@ -1,31 +1,39 @@
-const _ = require('lodash');
+const fs = require('fs').promises;
+const bluebird = require('bluebird');
+const { CHUNK_SIZE_SCHOOLING_REGISTRATION_DATA_PROCESSING } = require('../../infrastructure/constants');
```
Il n'y a qu'un seul endroit où on utilise les chunks, je propose donc de raccourcir en 
```suggestion
const { SCHOOLING_REGISTRATION_CHUNK_SIZE} = require('../../infrastructure/constants');
```

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,210 @@
 
-  if (_.isEmpty(parsedXmlDom)) {
-    return [];
+  const UAIFromSIECLE = await _extractUAI(path, encoding);
+
```
Expliciter qu'on stocke l'UAI dans externalID, et que l'orga en paramètre est celle liée à l'utilisateur
```suggestion
  const UAIFromSIECLE = await _extractUAI(path, encoding);
  const UAIFromUser = organization.externalId;
```

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,210 @@
-    return [];
+  const UAIFromSIECLE = await _extractUAI(path, encoding);
+
+  if (UAIFromSIECLE !== organization.externalId) {
```
```suggestion
  if (UAIFromSIECLE !== UAIFromUser) {
```

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,210 @@
+async function _readFirstLineFromFile(path) {
+  return new Promise((resolve, reject) => {
+    const readStream = fs.createReadStream(path);
+    const lineEnding = '\n';
```
Permet de souligner le fait que l'on recherche un caractère, pas une string
```suggestion
    const lineEndingCharacter = '\n';
```

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,210 @@
+        readStream.close();
+      }
+    })
+      .on('close', () => resolve(value.slice(value.charCodeAt(0) === BOM ? 1 : 0, position)))
```
Si on voulait être TRES explicite pour ceux qui ne savent pas que le premier caractère de début de fichier peut ne pas être significatif sur le contenu de la ligne ( et indique uniquement le format)

```suggestion
    const returnExtractedFirstLine = function() {
      const rawFirstLine = value;
      const lineStartsAt = rawFirstLine.charCodeAt(0) === BOM ? 1 : 0;
      const lineEndsAt = position;
      const firstLine = rawFirstLine.slice(lineStartsAt, lineEndsAt);
      resolve(firstLine);
    };
```

Ce qui donne le très explicite        
 ```suggestion
          .on('close', returnExtractedFirstLine)   
```

Pas de risque pour la perf, V8 va inliner tout ça

octo-topi replied:
D'ailleurs ici pour culture info, c'est quel cas sur les 3 ?
-     The byte order, or endianness, of the text stream in the cases of 16-bit and 32-bit encodings;
-     The fact that the text stream's encoding is Unicode, to a high level of confidence;
-     Which Unicode character encoding is used.

Si les fichiers SIECLE sont en ISO-8859-15, je penche pour `Which Unicode character encoding is used.`
https://en.wikipedia.org/wiki/Byte_order_mark

octo-topi commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,210 @@
+  return new Promise((resolve, reject) => {
+    const readStream = fs.createReadStream(path);
+    const lineEnding = '\n';
+    const BOM = 0xFEFF;
```
```suggestion
    const UTF32_BOM = 0xFEFF;
```
https://en.wikipedia.org/wiki/Byte_order_mark#Byte_order_marks_by_encoding

yahyayouness commented:
> Tu peux fixup le commit `[API] use buffer from file for siecle csv `file dans `[API] manage purge tmp siecle file after parsing ` car le nom n'est pas le bon (contenu réel = extraction d'un message dans une constante FILE_FORMAT_NOT_VALID)

Cette optimisation mérite un commit dédié voir une PR car c'est une optimisation du fichier siècle en csv qui consiste à utiliser le buffer de hapi. Ce dernier avait une portée plus globale et se libérait tardivement alors que la variable locale buffer rajoutée se libère très rapidement. 

Le contenu du commit est celui la:
`const buffer = await fs.readFile(payload.path);`
Une fois le buffer lu par la ligne d'après
`const csvSiecleParser = new SchoolingRegistrationParser(buffer, organizationId);`
Le buffer local est prêt à être libéré de suite.

J'ai juste enlevé un magic string dans ce commit.

lisequesnel commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,216 @@
 
-  return divisionStructureElement ? _getEitherElementValueOrNull(divisionStructureElement, 'CODE_STRUCTURE') : null;
+    saxParser.on('end', () => {
+      reject(new FileValidationError('UAI non trouvé dans le fichier'));
```
ici il faudrait mettre la même erreur que quand cela ne correspond pas : 
```suggestion
      reject(new FileValidationError(UAI_SIECLE_FILE_NOT_MATCH_ORGANIZATION_UAI));
```

lisequesnel commented on api/lib/domain/services/schooling-registrations-xml-service.js:
```diff
@@ -1,91 +1,216 @@
+  return new Promise(function(resolve, reject_) {
+    const saxParser = sax.createStream(true);
+    saxParser.on('error', () => {
+      reject(new FileValidationError('XML invalide'));
```
ici il faudrait mettre la même erreur que précédemment : 
```suggestion
      reject(new FileValidationError('Aucun élève n’a pu être importé depuis ce fichier. Vérifiez que le fichier est conforme.''));
```

lisequesnel commented on api/lib/domain/usecases/import-schooling-registrations-from-siecle.js:
```diff
@@ -1,31 +1,39 @@
 const SchoolingRegistrationParser = require('../../infrastructure/serializers/csv/schooling-registration-parser');
 
-module.exports = async function importSchoolingRegistrationsFromSIECLE({ organizationId, buffer, format, organizationRepository, schoolingRegistrationsXmlService, schoolingRegistrationRepository }) {
+const NO_SCHOOLING_REGISTRATIONS_FOUND = 'Aucune inscription d‘élève n‘a pu être importée depuis ce fichier. Vérifiez que le fichier est conforme.';
```
```suggestion
const NO_SCHOOLING_REGISTRATIONS_FOUND = 'Aucun élève n’a pu être importé depuis ce fichier. Vérifiez que le fichier est conforme.';
```

lisequesnel commented on api/tests/acceptance/application/organizations/organization-controller-import-schooling-registrations_test.js:
```diff
@@ -750,7 +750,7 @@ describe('Acceptance | Application | organization-controller-import-schooling-re
           expect(schoolingRegistrations).to.have.lengthOf(0);
           expect(response.statusCode).to.equal(422);
-          expect(response.result.errors[0].detail).to.equal('Aucune inscription d’élève n’a pu être importée depuis ce fichier. Vérifiez que le fichier est conforme.');
+          expect(response.result.errors[0].detail).to.equal('Aucune inscription d‘élève n‘a pu être importée depuis ce fichier. Vérifiez que le fichier est conforme.');
```
L'apostrophe est à l'envers
```suggestion
          expect(response.result.errors[0].detail).to.equal('Aucun élève n’a pu être importé depuis ce fichier. Vérifiez que le fichier est conforme.');
```
